Avoid unnecessary update of a map when updatedWith changes nothing.

hochgi opened the pull request:
a common usage is when one's want to update only when a key is not found. like:
```scala
m.updatedWith(key){
  case None => mkNewValue()
  case some => some
}
```
This commit aims to avoid the "some" case redundant overhead.

hochgi committed:
Avoid unnecessary update of a map when updatedWith changes nothing.

a common usage is when one's want to update only when a key is not found. like:

m.updatedWith(key){
  case None => mkNewValue()
  case some => some
}

This commit aims to avoid the "some" case redundant overhead.
As a bonus, I also rewrote without the redundant tuple allocation as per suggested in the comments.

OndrejSpanel commented on src/library/scala/collection/immutable/SortedMap.scala:
```diff
@@ -112,6 +112,8 @@ trait SortedMapOps[K, +V, +CC[X, +Y] <: Map[X, Y] with SortedMapOps[X, Y, CC, _]
     (previousValue, nextValue) match {
       case (None, None) => coll
       case (Some(_), None) => this.removed(key).coll
+      case (Some(v0: AnyRef), Some(v1: AnyRef)) if v0 eq v1 => coll // V may override equals, thus we can't use == on AnyRef
```
I do not understand this part. Why would `==` fail with overriden `equals`? The overriden `equals` would be called even when accessing through `AnyRef`, would not it?

hochgi replied:
If someone has a custom type, which answer `==` based on the custom equals, if we use `==` here it changes the behavior of that program.
I used `eq` to be sure this change is safe and doesn't change users code's behavior.

hochgi replied:
I don't have a good example (as I would avoid overriding equals), but let's say you have a type `Color`.
it has 2 implementing classes. `RGB` & `RGBA`.
You override equals such that you can compare `RGB` and `RGBA` based on the visible colors alone.

Now you want to update a map, and replace an `RGB` with the equivalent `RGBA`.

If we only check for `==`, we won't replace that value.

OndrejSpanel replied:
> if we use == here it changes the behavior of that program.

I am afraid I still do not understand. In what way would the behaviour be changed?

OndrejSpanel replied:
That sounds like having a `case class AB(a: Double, b: Double)` and comparing only on `a`. Such design would seem broken to me - two values which have some observable differences, yet compare as equal. I am afraid you are right such cases can exist and would behave differently with the optimization you introduce. I am not sure how careful one has to be to maintain functionality for such strangely (I would even say wrongly) designed types.

hochgi replied:
> > if we use == here it changes the behavior of that program.
> 
> I am afraid I still do not understand. In what way would the behaviour be changed?

I'll try to build upon my lousy `Color` example.
Given such `RGB` and `RGBA` values, we can think about something like:
```scala
type Pixel = (Int, Int)
def overrideWithAlpha(pixelToColor: Map[Pixel, Color], mask: List[(Pixel, Double)]): Map[Pixel, Color] = {
  mask.foldLeft(pixelToColor){ case (p2c, (pixel, alpha)) =>
    p2c.updateWith(pixel){
      case Some(c: Color) => Some(RGBA(c.red, c.green, c.blue, alpha))
      case none => none
    }
  }
}
```
in this case, since `c: Color` can be of type `RBG`, and thus only compare the visible red/green/blue colors with RGBA,
even though this is a real update, if we just use `==`, it yields `true`, thus we need `eq` to get `false`:
```scala
scala> c == RGBA(c.red, c.green, c.blue, alpha)
val res0: Boolean = true

scala> c eq RGBA(c.red, c.green, c.blue, alpha)
val res1: Boolean = false
```

hochgi replied:
> That sounds like having a `case class AB(a: Double, b: Double)` and comparing only on `a`. Such design would seem broken to me - two values which have some observable differences, yet compare as equal. I am afraid you are right such cases can exist and would behave differently with the optimization you introduce. I am not sure how careful one has to be to maintain functionality for such strangely (I would even say wrongly) designed types.

I agree.
But figured that being extra safe here cost almost nothing, and we gain the advantage of not breaking existing programs.
Users are usually aware that they shouldn't mess with `equals` on types that are used as keys, but here we operate on the value.

sjrd replied:
Unfortunately, those considerations are irrelevant, because it is in fact impossible to distinguish `AnyRef`s from `AnyVal`s at run-time. All values that are `eq` will be caught by this clause, *and* all values that are `==` will be caught by the next one.

This clause can be kept, as it is in fact valid for `AnyVal`s as well. The second clause needs to be removed, however, since it would also capture `AnyRef`s.

sjrd replied:
Also note that the rest of this file already uses the idiom `value1.asInstanceOf[AnyRef] eq value2.asInstanceOf[AnyRef]` for this kind of optimizations. We should use the same idiom here.

OndrejSpanel replied:
You are correct, of course, sorry for being blind. We are talking about replacing **values** with equal **values**, not about equality on **keys**.  It is quite possible someone is using such classes as values in the map, and it is important to preserve current behaviour for them.

hochgi replied:
@sjrd thanks, I wasn't aware it is impossible to distinguish at runtime.
Pushed a commit to change that.

som-snytt commented:
I see the tuple allocation is not eliminated, then I wondered if something clever were possible:
```
    nextValue match {
      case _: None.type => if (previousValue.isInstanceOf[Some[_]]) removed(key).coll else coll
      case Some(_: previousValue.type) => coll
      case Some(v) => updated(key, v)
    }
```
That avoids calling `equals` on `None` (some people like calling `Nil.equals` because it is a generous comparison), but Scala 2 doesn't use `eq` for comparison with `previousValue` even though it is a singleton pattern. (It uses `eq` for `V <: AnyRef`. Scala 3 is same behavior but no warning.)

hochgi commented:
@som-snytt I don't understand the second case. Should'nt we still check for `previousValue.get.asInstanceOf[AnyRef] eq nextValue.get.asInstanceOf[AnyRef]`?

som-snytt commented:
@hochgi I created an issue for clarification https://github.com/scala/bug/issues/12569

It looks like I wrote that code wrong, you're right it should be checking the value in the Option. Now I'll have to try it, though I'd be surprised if the behavior is different. I expected it to test `eq` for `case _: foo.type`. But if `foo` is an abstract type, what it does depends on whether `T <: AnyRef`.

hochgi commented:
@som-snytt before I submit another fix and trigger the build, would you prefer I'll do it in the style you wrote?
(gotta say, I don't see the appeal in writing `_: None.type` instead of just `None`. is it more performant?)

An alternative I find more to my liking (without the tuple allocation or redundant update):
```scala
val previousValue = this.get(key)
remappingFunction(previousValue) match {
  case None            => previousValue.fold(coll)(_ => coll.removed(key))
  case Some(nextValue) =>
    if (previousValue.exists(_.asInstanceOf[AnyRef] eq nextValue.asInstanceOf[AnyRef])) coll
    else coll.updated(key, nextValue)
```

som-snytt commented:
Sorry, I didn't mean to start a round of golf. :)  I don't have a strong opinion, except about the ticket I filed.

I like both your examples, and people have preferences. The minimal change you had is more conservative. It's your choice.

hochgi commented:
@som-snytt no worries, I enjoy this :)
Initially I meant for a minimal change,
but I don't mind being less conservative,
and new code does avoid tuple allocation as well :)

som-snytt commented:
Just a quick request to remove the at-notify in the commit message. Otherwise it will be noisy.

Good luck, I hope they like it! I hope it works on scala.js, etc.

OndrejSpanel reviewed (approved):
Being timid, esp. when doing changes in std lib, I guess I would prefer the more conservative change, but I agree this one has a potential of being more performant. That said, I think JVM is pretty good at eliminating local objects, therefore I doubt anyone would be able to show any provable impact on JVM. (I think it might be measurable on JS, I do not think local var elimination is yet that good there).

I think the new version achieves the desired goal, does not disrupt anything, is almost as well readable as the previous one and avoiding tuple allocation is just a small bonus.

som-snytt commented:
hey, I just used `updatedWith`!

My previous comment should add, You can squash the commits with `rebase -i` and then `push -f` to force push to the branch. That is normal procedure and will be an opportunity to take @ name off the commit message. Do not be overly concerned with inducing CI to run for just a commit.

som-snytt commented:
I don't know how many ways to ask, please do not say my name in a commit message. Do not say @ and do not follow that with som-snytt in a commit message. Thanks!

I don't want to see a notification that you pushed to your private repo, or that the commit was used in some other way.

hochgi commented:
Sorry, I will amend and remove the `@` from the commit.
I did a hasty squash, and forgot you asked.
apologies :pray:

NthPortal commented on src/library/scala/collection/immutable/SortedMap.scala:
```diff
@@ -108,11 +108,11 @@ trait SortedMapOps[K, +V, +CC[X, +Y] <: Map[X, Y] with SortedMapOps[X, Y, CC, _]
   override def updatedWith[V1 >: V](key: K)(remappingFunction: Option[V] => Option[V1]): CC[K, V1] = {
     // Implementation has been copied from `MapOps`
```
anyone know why the implementation is copied?

lrytz replied:
The commet above says

>   // We override these methods to fix their return type (which would be `Map` otherwise)

In `SortedMapOps[K, +V, +CC[X, Y] <: ..., +C <: ...] extends MapOps[K, V, Map, C]`, note how `CC` is not forwarded to `MapOps`.

NthPortal reviewed (approved):
could you maybe add a test that uses a custom `Map` implementation that counts calls to `updated`?

lrytz commented:
Sure, thank you :)
